### PR TITLE
Add suspense boundary around search page results

### LINKED ISSUE #59: [Task]: Update Search Page so Page Loads Statically and Results Are Pulled in Async
### Summary

See: https://github.com/HHS/simpler-grants-gov/issues/1804

### Acceptance criteria

_No response_

### PR DESCRIPTION
**Pull request by acouch**
## Summary
Fixes #59 

## Changes proposed
This makes the search page static and adds a suspense boundary for the data being fetched by the server. The data comes from the API and is called from 3 components:

* [`<SearchPaginationFetch />`](https://github.com/navapbc/simpler-grants-gov/pull/101/files#diff-9dbdda5096b97ad049cccea24c5a046581d26c151a6f94fcc32c05cb33ee9dee)
* [`<SearchResultsHeaderFetch />`](https://github.com/navapbc/simpler-grants-gov/pull/101/files#diff-14a084f66c050414cc2bbd0875256511630438971022073301bbfe91c4aa8cd1)
* [`<SearchResultsListFetch />`](https://github.com/navapbc/simpler-grants-gov/pull/101/files#diff-aabe6a7d19434a9b26199430bbcde5d31a0790aebc4cd844b922ac2fa1348dce)

This also simplifies the state model by pushing state changes directly to the browser query params and rerendering the changed items. This makes things a lot simpler and thus a lot of state management code is removed and there results list is no longer wrapped in a form and passing form refs between components. This is the recommended approach by next: https://nextjs.org/learn/dashboard-app/adding-search-and-pagination

There are several items that needed to be shared among the client components: the query, total results count, and total pages. These are wrapped in a `<QueryProvider />` that updates the state of these items. This was added so that if someone enters a query in the text box and the clicks a filter their query is not lost, so that the "N Opportunities" text doesn't need to be rerendered when paging or sorting, and so that the pager stays the same length when paging or sorting.

The data is fetched a couple of times in a duplicative fashion, however this follows [NextJS best practice](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#sharing-data-between-components) since the requests are cached.

The pager has been updated to reload only when there is a change in the page length. Because of an issue with the way the pager renders, it is unavailable while data is being updated:

image

This is because the Truss React component [switches between a link and a button as it renders](https://github.com/trussworks/react-uswds/blob/main/src/components/Pagination/Pagination.tsx#L42) and there isn't an option to supply query arguments, so if a user where to click it they would lose the query params. 

Overall this puts us on nice footing for the upcoming work using NextJS best practice.

### CODE REVIEW

File: frontend/src/app/services/search/FormDataService.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    return values;
+  }
+
+  get status(): Set<string> {
+    return this.getValuesByKeyPrefix("status-");
+  }
+
+  get fundingInstrument(): Set<string> {
+    // Pass 'true' to handle the special case of splitting 'funding-instrument-'after the second dash
+    return this.getValuesByKeyPrefix("funding-instrument-", true);
```
Comment (andycochran) [inline frontend/src/app/services/search/FormDataService.ts]: Should `funding-instrument-` be `funding_instrument-` (with an underscore) instead? I imagine this'll come up again. We might want to establish a standard for how we format our query parameters. Tho I'm not sure if there are downsides to underscores.

Comment (andycochran) [reply frontend/src/app/services/search/FormDataService.ts]: ~Is this not a concern for `search-sort-by`?~

Edit: I misunderstood this function and should've read more closely before commenting. You can disregard my underscore comment. It doesn't matter for query param presentation

File: frontend/src/app/services/search/searchfetcher/APISearchFetcher.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    super();
+    this.searchApi = new SearchOpportunityAPI();
+  }
+
+  async fetchOpportunities(
+    searchInputs: QueryParamData,
+  ): Promise<SearchAPIResponse> {
+    try {
+      // Keep commented in case we need to simulate a delay to test loaders
+      await new Promise((resolve) => setTimeout(resolve, 6250));
```
Comment (andycochran) [inline frontend/src/app/services/search/searchfetcher/APISearchFetcher.ts]: This should be commented out?

File: frontend/src/components/Footer.tsx [hunk 3]
```diff
@@ -84,7 +84,7 @@ const Footer = () => {
         <Grid row gap>
           <Grid tablet={{ col: 4 }} desktop={{ col: 6 }}>
             <Image
-              className="maxh-15 margin-bottom-2 tablet:margin-bottom-0"
+              className="margin-bottom-2 tablet:margin-bottom-0"
```
Comment (andycochran) [inline frontend/src/components/Footer.tsx]: Why was `. maxh-15` removed? It's there so that the logo doesn't grow too large on mobile. Larger breakpoints add a max size. But on mobile the image will just fill the allotted space and be HUGE. 

image

**Comment by rylew1** [comment]
Would be curious to see this diagram updated if you have time: https://drive.google.com/file/d/14iXrcg5EsEvnqqhUeVgR-5F_X8PASw4d/view?usp=sharing
image

File: frontend/src/hooks/useSearchParamUpdater.ts [hunk 4]
```diff
@@ -1,21 +1,20 @@
...
+  const router = useRouter();
+  const params = new URLSearchParams(searchParams);
 
-  // Singular string-type param updates include: search input, dropdown, and page numbers
-  // Multi/Set-type param updates include filters: Opportunity Status, Funding Instrument, Eligibility, Agency, Category
   const updateQueryParams = (
     queryParamValue: string | Set<string>,
     key: string,
+    queryTerm: string | null | undefined,
+    scroll = false,
```
Comment (andycochran) [inline frontend/src/hooks/useSearchParamUpdater.ts:22]: scroll is a query param?

File: frontend/src/app/services/search/FormDataService.ts (hunk 1 above)
Comment (acouch) [reply frontend/src/app/services/search/FormDataService.ts]: This file was removed actually.

File: frontend/src/app/services/search/searchfetcher/APISearchFetcher.ts (hunk 2 above)
Comment (acouch) [reply frontend/src/app/services/search/searchfetcher/APISearchFetcher.ts]: Yes, thanks.

File: frontend/src/components/Footer.tsx (hunk 3 above)
Comment (acouch) [reply frontend/src/components/Footer.tsx]: The max width creates an error in the console:

```
warn-once.js:16 Image with src "/_next/static/media/grants-gov-logo.7f89963e.png" has either width or height modified, but not the other. If you use CSS to change the size of your image, also include the styles 'width: "auto"' or 'height: "auto"' to maintain the aspect ratio
```

I removed the update and I created a separate ticket for it since it is out of scope for this ticket: https://github.com/navapbc/simpler-grants-gov/issues/165

File: frontend/src/hooks/useSearchParamUpdater.ts (hunk 4 above)
Comment (acouch) [reply frontend/src/hooks/useSearchParamUpdater.ts:22]: I documented the reason for this. It is implemented in the bottom pager so that when the user clicks the pager, they aren't left at the bottom of the screen when the results reload.

**Comment by acouch** [comment]
@rylew1 
diagram of components

As noted above, the workflow is:

1. request comes, static version of the page is rendered
2. search page params are args to search page
3. static client components are rendered [*on the server*](https://github.com/reactwg/server-components/discussions/4)
4. components within the suspense boundary ask for their data
5. data is fetched from the server (no change here) and results are streamed to the client
6. client sees results

When a user makes a change, nothing happens on the server. The query params are captured and pushed to update the new URL and the process from above is repeated.

**Comment by rylew1** [comment]
Is there a way to get the draw.io link or a larger picture of the updated diagram?

**Comment by acouch** [comment]
@rylew1 here is the file: https://drive.google.com/file/d/1ZPzAm4lZxfW5db9lt74mRCw73w-nbp4X/view?usp=sharing

File: frontend/src/app/[locale]/search/error.tsx [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
-import {
-  PaginationInfo,
-  SearchAPIResponse,
-} from "src/types/search/searchResponseTypes";
-
+import BetaAlert from "src/components/BetaAlert";
+import Breadcrumbs from "src/components/Breadcrumbs";
 import PageSEO from "src/components/PageSEO";
-import { QueryParamData } from "src/services/search/searchfetcher/SearchFetcher";
+import QueryProvider from "./QueryProvider";
```
Comment (btabaska) [inline frontend/src/app/[locale]/search/error.tsx]: Should be changed to import QueryProvider from "src/app/[locale]/search/QueryProvider"; to match expected import pattern

**Comment by btabaska** [comment]
General feedback- this PR is way too large and in the future we should consider creating team standards around PR size. [OpenSource.net](https://opensource.com/article/18/6/anatomy-perfect-pull-request) has a great article with research that shows that we should keep PRs between 10-250 lines of code to make reviewing more effective.

File: frontend/src/components/search/SearchBar.tsx [hunk 6]
```diff
@@ -1,21 +1,19 @@
 "use client";
-
 import { Icon } from "@trussworks/react-uswds";
-import { useSearchParamUpdater } from "../../hooks/useSearchParamUpdater";
-import { useState } from "react";
-import { sendGAEvent } from "@next/third-parties/google";
+import { QueryContext } from "../../app/[locale]/search/QueryProvider";
```
Comment (btabaska) [inline frontend/src/components/search/SearchBar.tsx]: Should be changed to import QueryProvider from "src/app/[locale]/search/QueryProvider"; to match expected import pattern

File: frontend/src/components/search/SearchBar.tsx [hunk 7]
```diff
@@ -1,21 +1,19 @@
 "use client";
-
 import { Icon } from "@trussworks/react-uswds";
-import { useSearchParamUpdater } from "../../hooks/useSearchParamUpdater";
-import { useState } from "react";
-import { sendGAEvent } from "@next/third-parties/google";
```
Comment (btabaska) [inline frontend/src/components/search/SearchBar.tsx:6]: Is there a reason that we are deprecating code changes for Google Analytics?

File: frontend/src/components/search/SearchBar.tsx [hunk 8]
```diff
@@ -1,21 +1,19 @@
...
 
-export default function SearchBar({ initialQueryParams }: SearchBarProps) {
-  const [inputValue, setInputValue] = useState<string>(initialQueryParams);
+export default function SearchBar({ query }: SearchBarProps) {
+  const { queryTerm, updateQueryTerm } = useContext(QueryContext);
   const { updateQueryParams } = useSearchParamUpdater();
 
   const handleSubmit = () => {
-    updateQueryParams(inputValue, "query");
-    sendGAEvent("event", "search", { search_term: inputValue });
```
Comment (btabaska) [inline frontend/src/components/search/SearchBar.tsx:18]: See above, if we deprecate the GAEvent it will break Google Analytics support.

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx [hunk 9]
```diff
@@ -23,40 +25,34 @@ export interface FilterOption {
...
+  id: string;
+  label: string;
+  value: string;
+  isChecked?: boolean;
+  children: FilterOption[];
+}
+
 interface SearchFilterAccordionProps {
-  initialFilterOptions: FilterOption[];
+  options: FilterOption[];
```
Comment (btabaska) [inline frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx]: What are your thoughts around "options" as the variable name? It might come across as confusing to developers working in the codebase in the future to have such a generic name. Since we have so many different types of options across the codebase, it might be more clear to keep the name "filterOptions" or something?

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterSection/SearchFilterSection.tsx [hunk 10]
```diff
@@ -1,65 +1,51 @@
 "use client";
-
-import { useEffect, useState } from "react";
-
-import { FilterOption } from "../SearchFilterAccordion";
+import { useState } from "react";
+import { FilterOptionWithChildren } from "../SearchFilterAccordion";
```
Comment (btabaska) [inline frontend/src/components/search/SearchFilterAccordion/SearchFilterSection/SearchFilterSection.tsx]: This should be updated to use the absolute import path from src

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterOptions.ts [hunk 11]
```diff
@@ -1,6 +1,215 @@
-import { FilterOption } from "../SearchFilterAccordion";
+import { FilterOption } from "./SearchFilterAccordion";
```
Comment (btabaska) [inline frontend/src/components/search/SearchFilterAccordion/SearchFilterOptions.ts]: This should be updated to use the absolute import path from src

File: frontend/src/components/search/SearchOpportunityStatus.tsx [hunk 12]
```diff
@@ -1,10 +1,8 @@
 "use client";
-
-import { useEffect, useState } from "react";
-
 import { Checkbox } from "@trussworks/react-uswds";
-import { useDebouncedCallback } from "use-debounce";
-import { useSearchParamUpdater } from "../../hooks/useSearchParamUpdater";
+import { QueryContext } from "../../app/[locale]/search/QueryProvider";
```
Comment (btabaska) [inline frontend/src/components/search/SearchOpportunityStatus.tsx]: This should be updated to use the absolute import path from src

File: frontend/src/components/search/SearchPagination.tsx [hunk 13]
```diff
@@ -1,76 +1,63 @@
...
 interface SearchPaginationProps {
-  showHiddenInput?: boolean; // Only one of the two SearchPagination should have this set
   page: number;
-  handlePageChange: (handlePage: number) => void; // managed in useSearchFormState
-  paginationRef?: React.RefObject<HTMLInputElement>; // managed in useSearchFormState
-  totalPages: number;
-  position: PaginationPosition;
-  searchResultsLength: number;
+  query: string | null | undefined;
+  total?: number | null;
```
Comment (btabaska) [inline frontend/src/components/search/SearchPagination.tsx:15]: What do you think about changing this property to be a little more descriptive like in the original codebase. total is harder to read on first glance compared to totalPages

File: frontend/src/components/search/SearchPagination.tsx [hunk 14]
```diff
@@ -1,76 +1,63 @@
...
-        <input
-          type="hidden"
-          name="currentPage"
-          ref={paginationRef}
-          value={page}
-          data-testid="hiddenCurrentPage"
-        />
-      )}
+    <div
+      style={{
```
Comment (btabaska) [inline frontend/src/components/search/SearchPagination.tsx:45]: Is this something that we can achieve by using our existing style or by creating a custom USWDS style? It is usually not best practice to style inline because it can cause styling creep over time that makes it hard to parse where CSS is being applied from.

File: frontend/src/components/search/SearchPaginationFetch.tsx [hunk 15]
```diff
@@ -0,0 +1,33 @@
+"use server";
+import { getSearchFetcher } from "src/services/search/searchfetcher/SearchFetcherUtil";
+import { QueryParamData } from "src/services/search/searchfetcher/SearchFetcher";
+import SearchPagination from "./SearchPagination";
```
Comment (btabaska) [inline frontend/src/components/search/SearchPaginationFetch.tsx]: This should be updated to use an absolute path

File: frontend/src/components/search/SearchResultsHeader.tsx [hunk 16]
```diff
@@ -1,30 +1,36 @@
...
+  queryTerm,
+  loading = false,
+}: {
+  sortby: string | null;
+  totalFetchedResults?: string;
+  queryTerm?: string | null | undefined;
+  loading?: boolean;
+}) {
+  const { totalResults } = useContext(QueryContext);
+  const total = totalFetchedResults || totalResults;
```
Comment (btabaska) [inline frontend/src/components/search/SearchResultsHeader.tsx:18]: Could we define this as totalResults to make it a bit more clear across the file?

File: frontend/src/components/search/SearchResultsHeader.tsx [hunk 17]
```diff
@@ -1,30 +1,36 @@
...
   return (
     <div className="grid-row">
-      <h2 className="tablet-lg:grid-col-fill margin-top-5 tablet-lg:margin-top-2 tablet-lg:margin-bottom-0">
-        {searchResultsLength} Opportunities
+      <h2
+        className="tablet-lg:grid-col-fill margin-top-5 tablet-lg:margin-top-2 tablet-lg:margin-bottom-0"
+        style={{ opacity: loading ? 0.5 : 1 }}
```
Comment (btabaska) [inline frontend/src/components/search/SearchResultsHeader.tsx]: Is this something that we can achieve by using our existing style or by creating a custom USWDS style? It is usually not best practice to style inline because it can cause styling creep over time that makes it hard to parse where CSS is being applied from.

File: frontend/src/components/search/SearchResultsHeaderFetch.tsx [hunk 18]
```diff
@@ -0,0 +1,26 @@
+"use server";
+import { getSearchFetcher } from "src/services/search/searchfetcher/SearchFetcherUtil";
+import { QueryParamData } from "src/services/search/searchfetcher/SearchFetcher";
+import SearchResultsHeader from "./SearchResultsHeader";
```
Comment (btabaska) [inline frontend/src/components/search/SearchResultsHeaderFetch.tsx:4]: This should be updated to use an absolute path.

File: frontend/src/components/search/SearchResultsListFetch.tsx [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  const searchFetcher = getSearchFetcher();
+  const searchResults = await searchFetcher.fetchOpportunities(searchParams);
+  const maxPaginationError = null;
+
+  if (searchResults.status_code !== 200) {
+    return <SearchErrorAlert />;
+  }
+
+  if (searchResults.data.length === 0) {
+    return (
```
Comment (btabaska) [inline frontend/src/components/search/SearchResultsListFetch.tsx:22]: These strings need to be migrated into locales so that translate will work on them. I don't think we want a pattern where some text is located there and some is located inline to the file.

File: frontend/src/components/search/SearchResultsListFetch.tsx [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    );
+  }
+
+  return (
+    <ul className="usa-list--unstyled">
+      {/* TODO #1485: show proper USWDS error  */}
+      {maxPaginationError && (
+        <h4>
+          {
+            "You''re trying to access opportunity results that are beyond the last page of data."
```
Comment (btabaska) [inline frontend/src/components/search/SearchResultsListFetch.tsx:41]: This string needs to be migrated into locales, we don't want inline text that will not be seen by translate and causes confusion on where we locate text.

File: frontend/src/hooks/useSearchParamUpdater.ts [hunk 21]
```diff
@@ -26,12 +31,20 @@ export function useSearchParamUpdater() {
...
+    if (queryTerm) {
+      params.set("query", queryTerm);
+    } else {
+      params.delete("query");
+    }
+    if (key !== "page") {
+      params.delete("page");
+    }
 
+    sendGAEvent("event", "search", { key: finalQueryParamValue });
```
Comment (btabaska) [inline frontend/src/hooks/useSearchParamUpdater.ts]: "key" needs to be changed either to "search_term" to match what we have in GA or it needs to be updated in GA to "key". GA won't pick up or understand what "key" means on its own. I'd recommend we change it to search_term here since "key" is generic and means a lot of different things.

File: frontend/src/app/[locale]/search/error.tsx (hunk 5 above)
Comment (acouch) [reply frontend/src/app/[locale]/search/error.tsx]: @btabaska if the file is in the same directory I'm not sure if it is better to use the full string or not. I created https://github.com/navapbc/simpler-grants-gov/issues/91 to decide on a standard. Can we punt to that ticket?

**Review by btabaska — CHANGES_REQUESTED** [review]
Lots of small nitpicks across the changes. It was hard to review this as a whole since the changes are so sweeping. We should likely break up changes like this into 5-6 smaller more focused PRs in the future to make the review process more effective. Great changes overall though, and running it locally this seems like a major improvement. Bravo!

**Comment by acouch** [comment]
@btabaska I agree but not sure how to break this up. The key items are to render the page statically, which wasn't possible w/o introducing the suspense items. Introducing the suspense items one at a time would take a ton of extra work. This takes advantage of the downtime.

File: frontend/src/app/[locale]/search/error.tsx (hunk 5 above)
Comment (btabaska) [reply frontend/src/app/[locale]/search/error.tsx]: Yep, any of these nitpicks are just suggestions. Feel free to not incorporate into final feedback

**Comment by btabaska** [comment]
> @btabaska I agree but not sure how to break this up. The key items are to render the page statically, which wasn't possible w/o introducing the suspense items. Introducing the suspense items one at a time would take a ton of extra work. This takes advantage of the downtime.

Totally fine, and agreed. I just wanted to point that out as a general concern. If we were in a live environment and not during downtime I might suggest a feature branch or something. However, in this instance it isn't a big deal.

File: frontend/src/components/search/SearchBar.tsx (hunk 7 above)
Comment (acouch) [reply frontend/src/components/search/SearchBar.tsx:6]: @btabaska this was moved to `updateQueryParams()`https://github.com/navapbc/simpler-grants-gov/pull/101/files/df65dbcd61af66cbbd532950994fcec69e9fee1f#diff-2b80310ac8bc745891bc29ca3d584cec3bd2984f0117b59aa2148625e9ab856dR43

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx (hunk 9 above)
Comment (acouch) [reply frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx]: @btabaska since the interface is `SearchFilterAccordionProps` it seems cleaner to me to have it as just `options` as it would follow that they are options for SearchFilterAccordion. I changed it because the "initial" didn't make sense anymore as these are not capturing state. It is great to keep in mind community devs and like `filterOptions` makes more sense for you so updating it.

File: frontend/src/components/search/SearchResultsHeader.tsx (hunk 16 above)
Comment (acouch) [reply frontend/src/components/search/SearchResultsHeader.tsx:18]: `totalResults` is used by the global QueryContext, so `total` here is just a place holder that is used in two lines below and is passed as `totalResults` to the sort component. `{total && <>{total} Opportunities</>}` reads better to me than `tempTotalResults` or `finalTotalResults` but am open to suggestions.

File: frontend/src/components/search/SearchResultsListFetch.tsx (hunk 19 above)
Comment (acouch) [reply frontend/src/components/search/SearchResultsListFetch.tsx:22]: @btabaska I agree, they were not translated originally so was trying to keep the scope of the PR down, even as large as it is. I created a follow up: https://github.com/navapbc/simpler-grants-gov/issues/167

File: frontend/src/hooks/useSearchParamUpdater.ts (hunk 21 above)
Comment (acouch) [reply frontend/src/hooks/useSearchParamUpdater.ts]: Updated, thanks!

File: frontend/src/components/search/SearchBar.tsx (hunk 8 above)
Comment (acouch) [reply frontend/src/components/search/SearchBar.tsx:18]: Thanks for noting. See comment above.

File: frontend/src/components/search/SearchPagination.tsx (hunk 13 above)
Comment (acouch) [reply frontend/src/components/search/SearchPagination.tsx:15]: This is a good point and another collision with the global queryContext. Maybe the context vars could be renamed to make it easier.

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterOptions.ts (hunk 11 above)
Comment (acouch) [reply frontend/src/components/search/SearchFilterAccordion/SearchFilterOptions.ts]: Updated, thx.

File: frontend/src/components/search/SearchOpportunityStatus.tsx (hunk 12 above)
Comment (acouch) [reply frontend/src/components/search/SearchOpportunityStatus.tsx]: Updated, thx.

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterSection/SearchFilterSection.tsx (hunk 10 above)
Comment (acouch) [reply frontend/src/components/search/SearchFilterAccordion/SearchFilterSection/SearchFilterSection.tsx]: Updated, thx.

File: frontend/src/app/[locale]/search/error.tsx (hunk 5 above)
Comment (acouch) [reply frontend/src/app/[locale]/search/error.tsx]: I updated anyway, we can figure out the convention in #91 and ideally enforce with a linter.

File: frontend/src/components/search/SearchResultsHeader.tsx (hunk 17 above)
Comment (acouch) [reply frontend/src/components/search/SearchResultsHeader.tsx]: Updated, thanks for noting.

**Review by btabaska — APPROVED** [review]
LGTM!
